### PR TITLE
Pkcs11 wrapper implementation

### PR DESCRIPTION
**Pull request by mykola-kobets-epam**

### CODE REVIEW

File: src/common/pkcs11/pkcs11.cpp [hunk 1]
```diff
@@ -0,0 +1,1038 @@
...
+        tmp.pValue = (void*)attr.mValue.Get();
+        tmp.ulValueLen = attr.mValue.Size();
+
+        dst.PushBack(tmp);
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error Merge(const Array<AttributeType>& types, Array<Array<uint8_t>>& values, Array<CK_ATTRIBUTE>& dst)
```
Comment (al1img) [inline src/common/pkcs11/pkcs11.cpp:168]: Merge what? Provide more meaningful name.

File: src/common/pkcs11/pkcs11.cpp [hunk 2]
```diff
@@ -0,0 +1,1038 @@
...
+        tmp.pValue = static_cast<void*>(values[i].Get());
+        tmp.ulValueLen = values[i].MaxSize();
+
+        dst.PushBack(tmp);
+    }
+
+    return AOS_ERROR_WRAP(ErrorEnum::eNone);
+}
+
+Error GetValues(Array<CK_ATTRIBUTE>& src, Array<Array<uint8_t>>& values)
```
Comment (al1img) [inline src/common/pkcs11/pkcs11.cpp:193]: ditto

File: src/common/pkcs11/pkcs11.cpp [hunk 3]
```diff
@@ -0,0 +1,1038 @@
...
+        }
+
+        auto tmp = Array<uint8_t>(static_cast<uint8_t*>(attr.pValue), attr.ulValueLen);
+        values.PushBack(tmp);
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Array<uint8_t> ConvertToByteArray(const String& val)
```
Comment (al1img) [inline src/common/pkcs11/pkcs11.cpp:213]: What happened here? Why not just `Convert` ;)?

File: src/common/pkcs11/pkcs11.cpp [hunk 4]
```diff
@@ -0,0 +1,1038 @@
...
+LibraryContext::LibraryContext(void* handle)
+    : mHandle(handle)
+{
+}
+
+Error LibraryContext::Init()
+{
+    CK_C_GetFunctionList getFuncList = reinterpret_cast<CK_C_GetFunctionList>(dlsym(mHandle, "C_GetFunctionList"));
+
+    const char* err;
```
Comment (al1img) [inline src/common/pkcs11/pkcs11.cpp:271]:
```c++
auto err = dlerror();
if (err != nullptr) {
}
```

File: include/aos/common/crypto.hpp [hunk 5]
```diff
@@ -333,6 +333,24 @@ inline void Encode(const Array<ObjectIdentifier>& src, Array<uint8_t>& asn1Value
     (void)asn1Value;
 }
 
+/**
+ * Encodes big integer in ASN1 format.
+ *
+ * @param number big integer.
+ * @param[out] asn1Value result ASN1 value.
```
Comment (mykola-kobets-epam) [inline include/aos/common/crypto.hpp:340]: [al1img](https://github.com/al1img) [Dec 21, 2023](https://github.com/aoscloud/aos_core_lib_cpp/pull/49#discussion_r1433702228)

Remove empty line

add [in] [out] for params
Author
@[mykola-kobets-epam](https://github.com/mykola-kobets-epam) mykola-kobets-epam [Dec 21, 2023](https://github.com/aoscloud/aos_core_lib_cpp/pull/49#discussion_r1433918605)

done

File: include/aos/common/config.hpp [hunk 6]
```diff
@@ -314,6 +307,34 @@
...
+ */
+#ifndef AOS_CONFIG_PKCS11_TOKEN_KEYS_COUNT
+#define AOS_CONFIG_PKCS11_TOKEN_KEYS_COUNT 20
+#endif
+
+/**
+ * Maximum number of PKCS11 libraries.
+ */
+#ifndef AOS_CONFIG_PKCS11_LIBRARIES_NUM
+#define AOS_CONFIG_PKCS11_LIBRARIES_NUM 1
```
Comment (mykola-kobets-epam) [inline include/aos/common/config.hpp:335]: [al1img]
set 1 as default. We usually use 1 library.
Author
@[mykola-kobets-epam](https://github.com/mykola-kobets-epam) mykola-kobets-epam [Dec 21, 2023](https://github.com/aoscloud/aos_core_lib_cpp/pull/49#discussion_r1433916483)

fixed

File: src/common/pkcs11/pkcs11.cpp [hunk 7]
```diff
@@ -0,0 +1,1077 @@
...
+ * Copyright (C) 2023 Renesas Electronics Corporation.
+ * Copyright (C) 2023 EPAM Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <dlfcn.h>
+#include <stdlib.h>
+#include <time.h>
+
```
Comment (mykola-kobets-epam) [inline src/common/pkcs11/pkcs11.cpp:11]: Header order:

    system headers section
    project headers section
    local headers

Author
@[mykola-kobets-epam](https://github.com/mykola-kobets-epam) mykola-kobets-epam [Dec 21, 2023](https://github.com/aoscloud/aos_core_lib_cpp/pull/49#discussion_r1433920539)

fixed

File: src/common/pkcs11/pkcs11.cpp [hunk 8]
```diff
@@ -0,0 +1,1077 @@
...
+            break;
+    }
+
+    dst.Resize(size);
+
+    return ErrorEnum::eNone;
+}
+
+template <size_t cSize>
+Error Convert(const CK_UTF8CHAR (&src)[cSize], String& dst)
```
Comment (mykola-kobets-epam) [inline src/common/pkcs11/pkcs11.cpp:53]: [al1img](https://github.com/al1img) [Dec 21, 2023](https://github.com/aoscloud/aos_core_lib_cpp/pull/49#discussion_r1433746256)

cSize
Author
@[mykola-kobets-epam](https://github.com/mykola-kobets-epam) mykola-kobets-epam [Dec 21, 2023](https://github.com/aoscloud/aos_core_lib_cpp/pull/49#discussion_r1433921504)

done

File: src/common/pkcs11/pkcs11.cpp [hunk 9]
```diff
@@ -0,0 +1,1077 @@
...
+{
+    dst.mMajor = src.major;
+    dst.mMinor = src.minor;
+}
+
+Error Convert(const CK_SLOT_INFO& src, SlotInfo& dst)
+{
+    Error err = Convert(src.manufacturerID, dst.mManufacturerID);
+    if (!err.IsNone()) {
+        return err;
```
Comment (mykola-kobets-epam) [inline src/common/pkcs11/pkcs11.cpp:81]: [al1img](https://github.com/al1img) [Dec 21, 2023](https://github.com/aoscloud/aos_core_lib_cpp/pull/49#discussion_r1433759457) •

    Do not use AOS_ERROR_WRAP in aos_core_lib_cpp.
    return err; here and all the cases below.

Author
@[mykola-kobets-epam](https://github.com/mykola-kobets-epam) mykola-kobets-epam [Dec 21, 2023](https://github.com/aoscloud/aos_core_lib_cpp/pull/49#discussion_r1433928775)

done

File: src/common/pkcs11/pkcs11.cpp [hunk 10]
```diff
@@ -0,0 +1,1077 @@
...
+        tmp.pValue = (void*)attr.mValue.Get();
+        tmp.ulValueLen = attr.mValue.Size();
+
+        dst.PushBack(tmp);
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error BuildAttributes(const Array<AttributeType>& types, Array<Array<uint8_t>>& values, Array<CK_ATTRIBUTE>& dst)
```
Comment (mykola-kobets-epam) [inline src/common/pkcs11/pkcs11.cpp:169]: [al1img](https://github.com/al1img) [Dec 21, 2023](https://github.com/aoscloud/aos_core_lib_cpp/pull/49#discussion_r1433761567)

Merge what? Provide more meaningful name.
Author
@[mykola-kobets-epam](https://github.com/mykola-kobets-epam) mykola-kobets-epam [Dec 21, 2023](https://github.com/aoscloud/aos_core_lib_cpp/pull/49#discussion_r1433933880)

CK_ATTRIBUTE consists of type and value
this method merges two distinct type and value arrays into a single CK_ATTRIBUTE array.

I will renames it to BuildAttributes.

File: src/common/pkcs11/pkcs11.cpp [hunk 11]
```diff
@@ -0,0 +1,1077 @@
...
+//
+
+LibraryContext::LibraryContext(void* handle)
+    : mHandle(handle)
+{
+}
+
+Error LibraryContext::Init()
+{
+    CK_C_GetFunctionList getFuncList = reinterpret_cast<CK_C_GetFunctionList>(dlsym(mHandle, "C_GetFunctionList"));
```
Comment (mykola-kobets-epam) [inline src/common/pkcs11/pkcs11.cpp:271]: [al1img]
auto err = dlerror();
if (err != nullptr) {
}

Author

fixed

File: src/common/pkcs11/pkcs11.cpp (hunk 2 above)
Comment (al1img) [reply src/common/pkcs11/pkcs11.cpp:195]: What about `GetAttributesValues`?

File: src/common/pkcs11/pkcs11.cpp [hunk 12]
```diff
@@ -0,0 +1,1047 @@
...
+Array<uint8_t> Convert(T& val)
+{
+    return {reinterpret_cast<uint8_t*>(&val), sizeof(val)};
+}
+
+/***********************************************************************************************************************
+ * Public
+ **********************************************************************************************************************/
+
+//
```
Comment (al1img) [inline src/common/pkcs11/pkcs11.cpp:230]: As you implement several classes in one cpp file, no need to divide the file into public and private part. So use header for each class instead:
```c++
/***********************************************************************************************************************
 * LibraryContext
 **********************************************************************************************************************/

...

/***********************************************************************************************************************
* SessionContext
***********************************************************************************************************************/
```
Put public methods before private.

File: src/common/pkcs11/pkcs11.cpp [hunk 13]
```diff
@@ -0,0 +1,1047 @@
...
+        }
+
+        auto tmp = Array<uint8_t>(static_cast<uint8_t*>(attr.pValue), attr.ulValueLen);
+        values.PushBack(tmp);
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Array<uint8_t> Convert(const String& val)
```
Comment (al1img) [inline src/common/pkcs11/pkcs11.cpp:215]: Too many converts also some of them almost same argument type it is hard to guess which one in which case to use. Rename all the convert methods in order to understand conversion details.

**Review by al1img — APPROVED** [review]

File: src/common/pkcs11/pkcs11.cpp [hunk 14]
```diff
@@ -0,0 +1,1059 @@
...
+        return ErrorEnum::eNoMemory;
+    }
+
+    dst.Clear();
+
+    for (const auto& attr : src) {
+        CK_ATTRIBUTE tmp;
+
+        tmp.type = attr.mType;
+        tmp.pValue = (void*)attr.mValue.Get();
```
Comment (MykolaSolyanko) [inline src/common/pkcs11/pkcs11.cpp:160]: can we use here `static_cast` ?

File: src/common/pkcs11/pkcs11.cpp [hunk 15]
```diff
@@ -0,0 +1,1059 @@
...
+    }
+
+    CK_UTF8CHAR pkcsLabel[cLabelLen];
+
+    auto err = ConvertToPKCS11String(label, pkcsLabel);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    CK_RV rv = mFunctionList->C_InitToken(slotID, (CK_UTF8CHAR_PTR)pin.CStr(), pin.Size(), pkcsLabel);
```
Comment (MykolaSolyanko) [inline src/common/pkcs11/pkcs11.cpp:313]: can we use here reinterprate_cast here ?

File: src/common/pkcs11/pkcs11.cpp [hunk 16]
```diff
@@ -0,0 +1,1059 @@
...
+
+Error SessionContext::Login(UserType userType, const String& pin)
+{
+    if (!mFunctionList || !mFunctionList->C_Login) {
+        LOG_ERR() << "C_Login failed. Library is not initialized.";
+
+        return ErrorEnum::eFailed;
+    }
+
+    CK_RV rv = mFunctionList->C_Login(mHandle, userType, (CK_UTF8CHAR_PTR)pin.CStr(), pin.Size());
```
Comment (MykolaSolyanko) [inline src/common/pkcs11/pkcs11.cpp:466]: ditto

File: src/common/pkcs11/pkcs11.cpp [hunk 17]
```diff
@@ -0,0 +1,1059 @@
...
+        return ErrorEnum::eFailed;
+    }
+
+    err = GetAttributesValues(pkcsAttributes, values);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    return ErrorEnum::eNone;
+}
```
Comment (MykolaSolyanko) [inline src/common/pkcs11/pkcs11.cpp:533]: return GetAttributesValues(pkcsAttributes, values);

File: src/common/pkcs11/pkcs11.cpp [hunk 18]
```diff
@@ -0,0 +1,1059 @@
...
+        }
+    } else {
+        LOG_ERR() << "C_CloseSession failed. Library is not initialized.";
+    }
+}
+
+Error SessionContext::FindObjectsInit(const Array<ObjectAttribute>& templ)
+{
+    if (!mFunctionList || !mFunctionList->C_FindObjectsInit) {
+        LOG_ERR() << "C_FindObjectsInit failed. Library is not initialized.";
```
Comment (MykolaSolyanko) [inline src/common/pkcs11/pkcs11.cpp:622]: please add empty line after log

File: src/common/pkcs11/pkcs11.cpp [hunk 19]
```diff
@@ -0,0 +1,1059 @@
...
+SessionContext::~SessionContext()
+{
+    if (mFunctionList && mFunctionList->C_CloseSession) {
+        CK_RV rv = mFunctionList->C_CloseSession(mHandle);
+        if (rv != CKR_OK) {
+            LOG_ERR() << "C_CloseSession failed. error = " << rv;
+        }
+    } else {
+        LOG_ERR() << "C_CloseSession failed. Library is not initialized.";
+    }
```
Comment (MykolaSolyanko) [inline src/common/pkcs11/pkcs11.cpp:616]: please use the same code style as above
```
if condition == error{
   return error
}

rest code
```

File: src/common/pkcs11/pkcs11.cpp [hunk 20]
```diff
@@ -0,0 +1,1059 @@
...
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    err = mSession.DestroyObject(key.GetPubHandle());
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    return ErrorEnum::eNone;
```
Comment (MykolaSolyanko) [inline src/common/pkcs11/pkcs11.cpp:976]: `return  mSession.DestroyObject(key.GetPubHandle());`

**Review by MykolaSolyanko — APPROVED** [review]

File: src/common/pkcs11/pkcs11.cpp (hunk 15 above)
Comment (mykola-kobets-epam) [reply src/common/pkcs11/pkcs11.cpp:313]: basically that's a combination of reinterpret & const cast, as in some cases even readonly value require non-const buffers. So I used c-style cast instead. 
Replaced with a function.

File: src/common/pkcs11/pkcs11.cpp (hunk 18 above)
Comment (mykola-kobets-epam) [reply src/common/pkcs11/pkcs11.cpp:622]: done

File: src/common/pkcs11/pkcs11.cpp (hunk 20 above)
Comment (mykola-kobets-epam) [reply src/common/pkcs11/pkcs11.cpp:976]: done

File: src/common/pkcs11/pkcs11.cpp (hunk 17 above)
Comment (mykola-kobets-epam) [reply src/common/pkcs11/pkcs11.cpp:533]: done

File: src/common/pkcs11/pkcs11.cpp (hunk 16 above)
Comment (mykola-kobets-epam) [reply src/common/pkcs11/pkcs11.cpp:466]: done

File: src/common/pkcs11/pkcs11.cpp [hunk 21]
```diff
@@ -0,0 +1,1059 @@
...
+
+        return nullptr;
+    }
+
+    auto res = MakeShared<LibraryContext>(&mAllocator, handle);
+    if (!res || !res->Init().IsNone()) {
+        return nullptr;
+    }
+
+    mLibraries.EmplaceBack(library, res);
```
[the remark this answers is not in the document]
Comment (mykola-kobets-epam) [reply src/common/pkcs11/pkcs11.cpp:787]: fixed

**Review by al1img — APPROVED** [review]

**Pull request merged**